File: sdk/rust/src/token/assessed_custom_fee.rs
use hedera_proto::services;

use crate::protobuf::{
    FromProtobuf,
    ToProtobuf,
};
use crate::{
    AccountId,
    TokenId,
};

/// A custom transfer fee that was assessed during the handling of a CryptoTransfer.
#[derive(Debug, Eq, PartialEq, Clone)]
#[cfg_attr(feature = "ffi", derive(serde::Serialize))]
#[cfg_attr(feature = "ffi", serde(rename_all = "camelCase"))]
pub struct AssessedCustomFee {
    /// The amount of currency charged to each payer.
    pub amount: i64,

    /// The currency `amount` is charged in, if `None` the fee is in HBar.
    pub token_id: Option<TokenId>,

    /// The account that receives the fees that were charged.
    pub fee_collector_account_id: Option<AccountId>,

    /// A list of all accounts that were charged this fee.
    pub payer_account_id_list: Vec<AccountId>,
}

impl AssessedCustomFee {
    /// Create a new `AssessedCustomFee` from protobuf-encoded `bytes`.
    ///
    /// # Errors
    /// - [`Error::FromProtobuf`](crate::Error::FromProtobuf) if decoding the bytes fails to produce a valid protobuf.
    /// - [`Error::FromProtobuf`](crate::Error::FromProtobuf) if decoding the protobuf fails.
    pub fn from_bytes(bytes: &[u8]) -> crate::Result<Self> {
        FromProtobuf::from_bytes(bytes)
    }

    /// Convert `self` to a protobuf-encoded [`Vec<u8>`].
    #[must_use]
    pub fn to_bytes(&self) -> Vec<u8> {
        ToProtobuf::to_bytes(self)
    }
}

impl FromProtobuf<services::AssessedCustomFee> for AssessedCustomFee {
    fn from_protobuf(pb: services::AssessedCustomFee) -> crate::Result<Self>
    where
        Self: Sized,
    {
        Ok(Self {
            amount: pb.amount,
            token_id: pb.token_id.map(TokenId::from_protobuf).transpose()?,
            fee_collector_account_id: FromProtobuf::from_protobuf(pb.fee_collector_account_id)?,
            payer_account_id_list: FromProtobuf::from_protobuf(pb.effective_payer_account_id)?,
        })
    }
}

impl ToProtobuf for AssessedCustomFee {
    type Protobuf = services::AssessedCustomFee;

    fn to_protobuf(&self) -> Self::Protobuf {
        services::AssessedCustomFee {
            amount: self.amount,
            token_id: self.token_id.to_protobuf(),
            fee_collector_account_id: self.fee_collector_account_id.to_protobuf(),
            effective_payer_account_id: self.payer_account_id_list.to_protobuf(),
        }
    }
}

File: sdk/swift/Sources/Hedera/AssessedCustomFee.swift
import CHedera
import Foundation
import HederaProtobufs

/// A custom transfer fee that was assessed during the handling of a ``TransferTransaction``.
public struct AssessedCustomFee: Equatable, Codable {
    /// The amount of currency charged to each payer.
    public let amount: Int64

    /// The currency `amount` is charged in, if `None` the fee is in HBar.
    public let tokenId: TokenId?

    /// The account that receives the fees that were charged.
    public let feeCollectorAccountId: AccountId?

    /// A list of all accounts that were charged this fee.
    public let payerAccountIdList: [AccountId]

    public static func fromBytes(_ bytes: Data) throws -> Self {
        try Self(fromProtobufBytes: bytes)
    }

    public func toBytes() -> Data {
        toProtobufBytes()
    }
}

extension AssessedCustomFee: TryProtobufCodable {
    internal typealias Protobuf = Proto_AssessedCustomFee

    internal init(fromProtobuf proto: Protobuf) throws {
        let tokenId = proto.hasTokenID ? proto.tokenID : nil
        let feeCollectorAccountId = proto.hasFeeCollectorAccountID ? proto.feeCollectorAccountID : nil

        self.init(
            amount: proto.amount,
            tokenId: .fromProtobuf(tokenId),
            feeCollectorAccountId: try .fromProtobuf(feeCollectorAccountId),
            payerAccountIdList: try .fromProtobuf(proto.effectivePayerAccountID)
        )
    }

    internal func toProtobuf() -> Protobuf {
        .with { proto in
            proto.amount = amount
            tokenId?.toProtobufInto(&proto.tokenID)
            feeCollectorAccountId?.toProtobufInto(&proto.feeCollectorAccountID)
            proto.effectivePayerAccountID = payerAccountIdList.toProtobuf()
        }
    }
}
